bridge/device/cloud/manager_test.go

```diff
@@ -68,7 +71,78 @@ func (r *resourceDataSync) copy() resourceData {
 	}
 }
 
+func getUnauthorizedError() status.Status {
+	msg := pool.NewMessage(context.Background())
+	msg.SetCode(codes.Unauthorized)
+	return status.Errorf(msg, "unauthorized")
+}
+
 // device is restarted with an imported configuration with valid cloud credentials
+func TestManagerDeviceBecomesUnauthorized(t *testing.T) {
+	ch := mockCoapGW.NewCoapHandlerWithCounter(3600)
+	customHandler := mockCoapGW.NewCustomHandler(ch)
+	makeHandler := func(*mockCoapGWService.Service, ...mockCoapGWService.Option) mockCoapGWService.ServiceHandler {
+		return customHandler
+	}
+	coapShutdown := mockCoapGW.New(t, makeHandler, func(mockCoapGWService.ServiceHandler) {
+		h := ch
+		fmt.Printf("%+v\n", h.CallCounter.Data)
+		// d1 -> signup + signin + publish
+		// d2 -> should use the stored credentials to skip signup and only do sign in + publish
+		require.Equal(t, 1, h.CallCounter.Data[mockCoapGW.SignUpKey])
+		require.Equal(t, 1, h.CallCounter.Data[mockCoapGW.SignInKey])
+		require.Equal(t, 1, h.CallCounter.Data[mockCoapGW.PublishKey])
+		require.Equal(t, 0, h.CallCounter.Data[mockCoapGW.UnpublishKey])
+		require.Equal(t, 0, h.CallCounter.Data[mockCoapGW.RefreshTokenKey])
+	})
+	defer coapShutdown()
+
+	s1 := bridgeTest.NewBridgeService(t)
+	t.Cleanup(func() {
+		_ = s1.Shutdown()
+	})
+	deviceID := uuid.New().String()
+	d1 := bridgeTest.NewBridgedDevice(t, s1, deviceID, true, false, device.WithCloudOptions(cloud.WithTickInterval(time.Second)))
+	s1Shutdown := bridgeTest.RunBridgeService(s1)
+	t.Cleanup(func() {
+		_ = s1Shutdown()
+	})
+
+	c, err := testClient.NewTestSecureClientWithBridgeSupport()
+	require.NoError(t, err)
+	defer func() {
+		errC := c.Close(context.Background())
+		require.NoError(t, errC)
+	}()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+	err = c.OnboardDevice(ctx, deviceID, "authorizationProvider", "coaps+tcp://"+mockCoapGW.COAP_GW_HOST, "authorizationCode", test.CloudSID())
+	require.NoError(t, err)
+
+	// wait for sign in
+	require.Equal(t, 1, ch.WaitForSignIn(time.Second*20))
+
+	// wait for publish
+	require.Equal(t, 1, ch.WaitForPublish(time.Second*20))
+
+	customHandler.SetSignIn(func(ocfCloud.CoapSignInRequest) (ocfCloud.CoapSignInResponse, error) {
+		return ocfCloud.CoapSignInResponse{}, getUnauthorizedError()
+	})
+	customHandler.SetRefreshToken(func(ocfCloud.CoapRefreshTokenRequest) (ocfCloud.CoapRefreshTokenResponse, error) {
+		return ocfCloud.CoapRefreshTokenResponse{}, getUnauthorizedError()
+	})
+
+	d1.GetCloudManager().Reconnect()
+	for {
```

Comment:
I'd add a counter here, to avoid an infinite loop in case of a failure

Reply:
added